faiss_vector_posting.go

```diff
@@ -318,23 +348,61 @@ func (sb *SegmentBase) InterpretVectorIndex(field string, except *roaring.Bitmap
 					return rv, nil
 				}
 
-				scores, ids, err := vecIndex.SearchWithoutIDs(qVector, k, vectorIDsToExclude, params)
+				var scores []float32
+				var ids []int64
+				var err error
+
+				scores, ids, err = vecIndex.SearchWithoutIDs(qVector, k,
+					vectorIDsToExclude, params)
 				if err != nil {
 					return nil, err
 				}
-				// for every similar vector returned by the Search() API, add the corresponding
-				// docID and the score to the newly created vecPostingsList
-				for i := 0; i < len(ids); i++ {
-					vecID := ids[i]
-					// Checking if it's present in the vecDocIDMap.
-					// If -1 is returned as an ID(insufficient vectors), this will ensure
-					// it isn't added to the final postings list.
-					if docID, ok := vecDocIDMap[vecID]; ok {
-						code := getVectorCode(docID, scores[i])
-						rv.postings.Add(uint64(code))
+
+				addIDsToPostingsList(rv, ids, scores)
+
+				return rv, nil
+			},
+			searchWithFilter: func(qVector []float32, k int64,
+				eligibleDocIDs []uint64, params json.RawMessage) (
+				segment.VecPostingsList, error) {
+				// 1. returned postings list (of type PostingsList) has two types of information - docNum and its score.
+				// 2. both the values can be represented using roaring bitmaps.
+				// 3. the Iterator (of type PostingsIterator) returned would operate in terms of VecPostings.
+				// 4. VecPostings would just have the docNum and the score. Every call of Next()
+				//    and Advance just returns the next VecPostings. The caller would do a vp.Number()
+				//    and the Score() to get the corresponding values
+				rv := &VecPostingsList{
+					except:   nil, // todo: handle the except bitmap within postings iterator.
+					postings: roaring64.New(),
+				}
+
+				if vecIndex == nil || vecIndex.D() != len(qVector) {
+					// vector index not found or dimensionality mismatched
+					return rv, nil
+				}
+
+				var scores []float32
+				var ids []int64
+				var err error
+
+				if len(eligibleDocIDs) > 0 {
+					// Non-zero documents eligible per the filter query.
+
+					// vector IDs corresponding to the local doc numbers to be
+					// considered for the search
+					vectorIDsToInclude := make([]int64, len(eligibleDocIDs))
+					for i, id := range eligibleDocIDs {
+						vectorIDsToInclude[int64(i)] = docVecIDMap[uint32(id)]
+					}
+
+					scores, ids, err = vecIndex.SearchWithIDs(qVector, k,
+						vectorIDsToInclude, params)
```

Comment:
Shouldn't you be removing `vectorIDsToExclude` from `vectorIDsToInclude` before this or are we certain there never really is going to be an overlap there because of the pre-filter?

Reply:
As per my understanding, deleted results aren't returned as part of the initial filter search, whose results form the basis of the vector include list.